core/types/subtyping.cc

```diff
@@ -404,28 +404,35 @@ TypePtr Types::lub(const GlobalState &gs, const TypePtr &t1, const TypePtr &t2)
             ENFORCE(result != nullptr);
             return result;
         } else {
-            bool allowProxyInLub = isa_type<TupleType>(t1) || isa_type<ShapeType>(t1);
             // only 1st is proxy
-            TypePtr und = t1.underlying(gs);
-            if (isSubType(gs, und, t2)) {
-                return t2;
-            } else if (allowProxyInLub) {
-                return OrType::make_shared(t1, t2);
-            } else {
-                return lub(gs, t2, und);
-            }
+            return lub(gs, t2, t1);
         }
     } else if (is_proxy_type(t2)) {
         // only 2nd is proxy
         bool allowProxyInLub = isa_type<TupleType>(t2) || isa_type<ShapeType>(t2);
-        // only 1st is proxy
-        TypePtr und = t2.underlying(gs);
-        if (isSubType(gs, und, t1)) {
+        TypePtr und2 = t2.underlying(gs);
+        if (isSubType(gs, und2, t1)) {
             return t1;
         } else if (allowProxyInLub) {
-            return OrType::make_shared(t1, t2);
+            auto *orType1 = cast_type<OrType>(t1);
+            if (orType1 != nullptr) {
+                auto t2LubT1Left = lub(gs, t2, orType1->right);
+                if (t2LubT1Left == orType1->right) {
+                    return t1;
+                } else if (t2LubT1Left == t2) {
+                    return OrType::make_shared(orType1->left, t2);
+                }
+                auto t2LubT1Right = lub(gs, t2, orType1->right);
+                if (t2LubT1Right == orType1->right) {
+                    return t1;
+                } else if (t2LubT1Right == t2) {
+                    return OrType::make_shared(orType1->left, t2);
+                }
+            } else {
+                return OrType::make_shared(t1, t2);
+            }
         } else {
-            return lub(gs, t1, und);
+            return OrType::make_shared(t1, t2);
```

Comment:
I have to rethink this entire `// only 2nd is proxy` case to see if it applies specifically to this `T.any(Shape, Shape)` PR. I had a number of changes in the area of Tuples, Shapes, Hashes and tried best to split them up into small changes associated with specific issues.

This block would change the result of 'proxy / non-proxy' interactions making `T.any(...)` e.g. for 'Tuple / non-Tuple/Shape' as well as 'Shape / non-Shape/Tuple'.

It could be a separate PR for the case where a Shape (or Tuple) is being Or-combined (lub'ed) with a Shape (or Tuple) and one is a supertype of the other.

Reply:
Extracted into friend `lubProxyNonProxy()`